File: reference/haskell/src/TestVectors.hs

-- | Generate test vectors to compare with other implementations

module TestVectors where

--------------------------------------------------------------------------------

import Control.Monad

import Data.Word
import qualified Data.ByteString as B

import Poseidon2.Merkle
import Poseidon2.Sponge
import Slot

import ZK.Algebra.Curves.BN128.Fr.Mont (Fr)

--------------------------------------------------------------------------------

allTestVectors = do
  testVectorsSponge 
  testVectorsHash
  testVectorsMerkle

--------------------------------------------------------------------------------

testVectorsSponge :: IO ()
testVectorsSponge = do
  putStrLn ""
  putStrLn "test vectors for sponge of field elements with rate=1"
  putStrLn "-----------------------------------------------------"
  forM_ [0..8] $ \n -> do
    let input = map fromIntegral [1..n] :: [Fr]
    putStrLn $ "hash of [1.." ++ show n ++ "] :: [Fr] =  " ++ show (sponge1 input)

  putStrLn ""
  putStrLn "test vectors for sponge of field elements with rate=2"
  putStrLn "-----------------------------------------------------"
  forM_ [0..8] $ \n -> do
    let input = map fromIntegral [1..n] :: [Fr]
    putStrLn $ "hash of [1.." ++ show n ++ "] :: [Fr] =  " ++ show (sponge2 input)

--------------------------------------------------------------------------------

testVectorsHash :: IO ()
testVectorsHash = do

  putStrLn ""
  putStrLn "test vectors for hash (padded sponge with rate=2) of bytes"
  putStrLn "----------------------------------------------------------"
  forM_ [0..80] $ \n -> do
    let input = map fromIntegral [1..n] :: [Word8]
    let bs    = B.pack input
    putStrLn $ "hash of [1.." ++ show n ++ "] :: [Byte] =  " ++ show (hashCell_ bs)

--------------------------------------------------------------------------------

testVectorsMerkle :: IO ()
testVectorsMerkle = do
  putStrLn ""
  putStrLn "test vectors for Merkle roots of field elements"
  putStrLn "-----------------------------------------------"
  forM_ [1..40] $ \n -> do
    let input = map fromIntegral [1..n] :: [Fr]
    putStrLn $ "Merkle root of [1.." ++ show n ++ "] :: [Fr]  =  " ++ show (calcMerkleRoot input)

  putStrLn ""
  putStrLn "test vectors for Merkle roots of sequence of bytes"
  putStrLn "--------------------------------------------------"
  forM_ [0..80] $ \n -> do
    let input = map fromIntegral [1..n] :: [Word8]
    let bs    = B.pack input
    let flds  = cellDataToFieldElements (CellData bs)
    putStrLn $ "Merkle root of [1.." ++ show n ++ "] :: [Byte]  =  " ++ show (calcMerkleRoot flds)

--------------------------------------------------------------------------------

File: reference/nim/testvectors/src/testvectors.nim

import sugar

import std/math
import std/sequtils

#import constantine/math/arithmetic
import constantine/math/io/io_fields
import constantine/math/io/io_bigints
#import constantine/serialization/codecs

import poseidon2/types
import poseidon2/io
#import poseidon2/compress
import poseidon2/merkle
import poseidon2/sponge

#-------------------------------------------------------------------------------

proc testVectorsSponge() = 
  echo( "" )
  echo( "NIM | test vectors for sponge of field elements with rate=1" )
  echo( "-----------------------------------------------------------" )
  for n in 0..8:
    let input : seq[F] = collect( newSeq , (for i in 1..n: toF(i)) )
    let hash = Sponge.digest(input, rate = 1)
    echo( "hash of [1.." & ($n) & "] : seq[F] =  " & toDecimal(hash) )

  echo( "" )
  echo( "NIM | test vectors for sponge of field elements with rate=2" )
  echo( "-----------------------------------------------------------" )
  for n in 0..8:
    let input : seq[F] = collect( newSeq , (for i in 1..n: toF(i)) )
    let hash = Sponge.digest(input, rate = 2)
    echo( "hash of [1.." & ($n) & "] : seq[F] =  " & toDecimal(hash) )

#-------------------------------------------------------------------------------

proc testVectorsHash() = 
  echo ""
  echo "NIM | test vectors for hash (padded sponge with rate=2) of bytes"
  echo "----------------------------------------------------------------"
  for n in 0..80:
    let input : seq[byte] = collect( newSeq , (for i in 1..n: byte(i)) )
    let hash = Sponge.digest(input, rate=2)
    echo( "hash of [1.." & ($n) & "] : seq[byte] =  " & toDecimal(hash) )

#-------------------------------------------------------------------------------

proc testVectorsMerkle() =
  echo ""
  echo "NIM | test vectors for Merkle roots of field elements"
  echo "-----------------------------------------------------"
  for n in 1..40:
    let input : seq[F] = collect( newSeq , (for i in 1..n: toF(i)) )
    let root = Merkle.digest(input)
    echo( "Merkle root of [1.." & ($n) & "] : seq[F] =  " & toDecimal(root) )

  echo ""
  echo "NIM | test vectors for Merkle roots of sequence of bytes"
  echo "--------------------------------------------------------"
  for n in 0..80:
    let input : seq[byte] = collect( newSeq , (for i in 1..n: byte(i)) )
    let root = Merkle.digest(input)
    echo( "Merkle root of [1.." & ($n) & "] : seq[byte] =  " & toDecimal(root) )

#-------------------------------------------------------------------------------

when isMainModule:
  testVectorsSponge()
  testVectorsHash()
  testVectorsMerkle()

#-------------------------------------------------------------------------------
